File: conformance/golden-file-worldline/v1/scenario_oracle.py
"""Independent Golden File Worldline mutation and scenario laws."""

from __future__ import annotations

from corpus_protocol import (
    MAX_MUTATION_VALUE_BYTES,
    MAX_SOURCE_BYTES,
    ROOT,
    case_name,
    decimal,
    decoded_hex,
    fail,
    rows,
    unique,
)
from identity_oracle import (
    ALGORITHM,
    BINARY_LENGTH,
    ID_MAGIC,
    VERSION,
    IdentityFixture,
    digest,
    expected_text,
)

STEP_COLUMNS = [
    "step",
    "operation",
    "input_case",
    "identity_case",
    "expected_outcome",
]
MUTATION_COLUMNS = [
    "case",
    "target_kind",
    "target_case",
    "operation",
    "offset",
    "value_hex",
    "expected_outcome",
]
STEP_OUTCOMES = {
    "identify": "keep.identity.identified",
    "admit-exact": "keep.content.admitted",
    "read-exact": "keep.content.exact",
    "verify-claimed-content": "keep.content.mismatch",
    "read-absent": "keep.content.absent",
}
REQUIRED_STEP_OPERATIONS = [
    "identify",
    "admit-exact",
    "read-exact",
    "identify",
    "admit-exact",
    "read-exact",
    "read-exact",
    "verify-claimed-content",
    "read-exact",
    "read-absent",
]
REQUIRED_MUTATION_OPERATIONS = {
    ("content", "xor-byte"),
    ("content", "truncate"),
    ("content", "append"),
    ("identity-binary", "xor-byte"),
    ("identity-binary", "truncate"),
    ("identity-binary", "append"),
    ("identity-binary", "set-u8"),
    ("identity-binary", "set-u16-be"),
}


def mutate(
    target: bytes,
    operation: str,
    offset: int,
    value_field: str,
    name: str,
) -> bytes:
    changed = bytearray(target)
    if operation == "truncate":
        if value_field != "-" or offset >= len(changed):
            fail(f"{name}: invalid truncation")
        del changed[offset:]
    elif operation == "append":
        value = decoded_hex(
            value_field,
            f"{name} value",
            MAX_MUTATION_VALUE_BYTES,
        )
        if offset != len(changed):
            fail(f"{name}: append offset does not equal target length")
        changed.extend(value)
    else:
        width = {
            "xor-byte": 1,
            "set-u8": 1,
            "set-u16-be": 2,
        }.get(operation)
        if width is None:
            fail(f"{name}: unknown mutation operation {operation!r}")
        value = decoded_hex(value_field, f"{name} value", width)
        if len(value) != width or offset + width > len(changed):
            fail(f"{name}: mutation width or offset is invalid")
        if operation == "xor-byte":
            changed[offset] ^= value[0]
        else:
            changed[offset : offset + width] = value
    result = bytes(changed)
    if (
        result == target
        or len(result) > MAX_SOURCE_BYTES + MAX_MUTATION_VALUE_BYTES
    ):
        fail(f"{name}: mutation is a no-op or exceeds its bound")
    return result


def binary_outcomes(encoded: bytes) -> set[str]:
    if len(encoded) < BINARY_LENGTH:
        return {"keep.identity.truncated"}
    if len(encoded) > BINARY_LENGTH:
        return {"keep.identity.trailing_data"}
    if encoded[:16] != ID_MAGIC:
        return {"keep.identity.invalid_magic"}
    if int.from_bytes(encoded[16:18], "big") != VERSION:
        return {"keep.identity.unsupported_version"}
    if encoded[18] != ALGORITHM:
        return {"keep.identity.unsupported_algorithm"}
    return {"keep.identity.different_supported_identity"}


def check_mutations(
    fixtures: dict[str, IdentityFixture],
) -> None:
    mutation_rows = rows(
        ROOT / "mutations.tsv",
        "# keep.golden-file-worldline.mutations/v1",
        MUTATION_COLUMNS,
    )
    seen: set[str] = set()
    covered: set[tuple[str, str]] = set()
    for row in mutation_rows:
        name = case_name(row["case"], "mutations.tsv")
        unique(name, seen, "mutations.tsv")
        target_kind = row["target_kind"]
        if target_kind not in {"content", "identity-binary"}:
            fail(f"{name}: unknown mutation target kind")
        fixture = fixtures.get(row["target_case"])
        if fixture is None:
            fail(f"{name}: mutation target case is absent")
        target = (
            fixture.content
            if target_kind == "content"
            else fixture.canonical_binary
        )
        offset = decimal(row["offset"], f"{name} offset", len(target))
        changed = mutate(
            target,
            row["operation"],
            offset,
            row["value_hex"],
            name,
        )
        if target_kind == "content":
            observed = expected_text(len(changed), digest(changed))
            if (
                row["expected_outcome"] != "keep.content.mismatch"
                or observed == fixture.canonical_text
            ):
                fail(
                    f"{name}: content mutation does not prove mismatch"
                )
        elif row["expected_outcome"] not in binary_outcomes(changed):
            fail(f"{name}: binary mutation outcome is incorrect")
        covered.add((target_kind, row["operation"]))
    if not REQUIRED_MUTATION_OPERATIONS.issubset(covered):
        fail("mutations.tsv: required v1 mutation coverage is absent")


def check_steps(fixtures: dict[str, IdentityFixture]) -> None:
    step_rows = rows(
        ROOT / "steps.tsv",
        "# keep.golden-file-worldline.steps/v1",
        STEP_COLUMNS,
    )
    admitted: set[bytes] = set()
    operations: list[str] = []
    for expected_number, row in enumerate(step_rows, start=1):
        number = decimal(
            row["step"],
            "scenario step",
            len(step_rows),
        )
        if number != expected_number:
            fail(
                "steps.tsv: step numbers are not canonical and contiguous"
            )
        operation = row["operation"]
        outcome = STEP_OUTCOMES.get(operation)
        if outcome is None or row["expected_outcome"] != outcome:
            fail(f"steps.tsv:{number}: operation outcome is invalid")
        identity = fixtures.get(row["identity_case"])
        if identity is None:
            fail(f"steps.tsv:{number}: identity case is absent")
        if operation == "read-absent":
            if (
                row["input_case"] != "-"
                or identity.canonical_binary in admitted
            ):
                fail(f"steps.tsv:{number}: absent read is not absent")
        else:
            source = fixtures.get(row["input_case"])
            if source is None:
                fail(f"steps.tsv:{number}: input case is absent")
            same_identity = (
                source.canonical_binary == identity.canonical_binary
            )
            if operation == "verify-claimed-content" and same_identity:
                fail(f"steps.tsv:{number}: mismatch uses matching content")
            if (
                operation != "verify-claimed-content"
                and not same_identity
            ):
                fail(
                    f"steps.tsv:{number}: exact operation substitutes "
                    "identity"
                )
            if operation == "admit-exact":
                admitted.add(identity.canonical_binary)
            if (
                operation == "read-exact"
                and identity.canonical_binary not in admitted
            ):
                fail(
                    f"steps.tsv:{number}: exact read precedes admission"
                )
        operations.append(operation)
    if operations != REQUIRED_STEP_OPERATIONS:
        fail(
            "steps.tsv: ordered Golden File Worldline v1 operations moved"
        )

File: xtask/src/golden_file_worldline/scenario_oracle.rs
use std::collections::{BTreeMap, BTreeSet};

use super::canonical_value::decimal;
use super::identity_oracle::IdentityFixture;
use super::{Corpus, GoldenError};

const STEP_COLUMNS: [&str; 5] = [
    "step",
    "operation",
    "input_case",
    "identity_case",
    "expected_outcome",
];
const REQUIRED_STEP_OPERATIONS: [&str; 10] = [
    "identify",
    "admit-exact",
    "read-exact",
    "identify",
    "admit-exact",
    "read-exact",
    "read-exact",
    "verify-claimed-content",
    "read-exact",
    "read-absent",
];
pub(super) fn check_steps(
    corpus: &Corpus,
    fixtures: &BTreeMap<String, IdentityFixture>,
) -> Result<(), GoldenError> {
    let rows = corpus.rows(
        "steps.tsv",
        "# keep.golden-file-worldline.steps/v1",
        &STEP_COLUMNS,
    )?;
    let maximum = u64::try_from(rows.len()).map_err(|source| {
        GoldenError::violation(format!(
            "scenario row count cannot be represented: {source}"
        ))
    })?;
    let mut admitted = BTreeSet::new();
    let mut operations = Vec::new();
    for (offset, row) in rows.into_iter().enumerate() {
        let expected_number = offset
            .checked_add(1)
            .ok_or_else(|| GoldenError::violation("scenario step number overflow"))?;
        let number = decimal(row.field("step")?, "scenario step", maximum)?;
        let expected_u64 = u64::try_from(expected_number).map_err(|source| {
            GoldenError::violation(format!("scenario step cannot be represented: {source}"))
        })?;
        if number != expected_u64 {
            return Err(GoldenError::violation(
                "steps.tsv: step numbers are not canonical and contiguous",
            ));
        }
        let operation = row.field("operation")?;
        let expected_outcome = step_outcome(operation).ok_or_else(|| {
            GoldenError::violation(format!("steps.tsv:{number}: operation outcome is invalid"))
        })?;
        if row.field("expected_outcome")? != expected_outcome {
            return Err(GoldenError::violation(format!(
                "steps.tsv:{number}: operation outcome is invalid"
            )));
        }
        if operation == "read-absent" {
            let identity = fixtures.get(row.field("identity_case")?).ok_or_else(|| {
                GoldenError::violation(format!("steps.tsv:{number}: identity case is absent"))
            })?;
            if row.field("input_case")? != "-" || admitted.contains(&identity.canonical_binary) {
                return Err(GoldenError::violation(format!(
                    "steps.tsv:{number}: absent read is not absent"
                )));
            }
        } else {
            check_exact_step(&row, fixtures, operation, number, &mut admitted)?;
        }
        operations.push(operation.to_owned());
    }
    if operations
        .iter()
        .map(String::as_str)
        .eq(REQUIRED_STEP_OPERATIONS)
    {
        Ok(())
    } else {
        Err(GoldenError::violation(
            "steps.tsv: ordered Golden File Worldline v1 operations moved",
        ))
    }
}

fn check_exact_step(
    row: &super::corpus_protocol::TableRow,
    fixtures: &BTreeMap<String, IdentityFixture>,
    operation: &str,
    number: u64,
    admitted: &mut BTreeSet<Vec<u8>>,
) -> Result<(), GoldenError> {
    let identity = fixtures.get(row.field("identity_case")?).ok_or_else(|| {
        GoldenError::violation(format!("steps.tsv:{number}: identity case is absent"))
    })?;
    let source = fixtures.get(row.field("input_case")?).ok_or_else(|| {
        GoldenError::violation(format!("steps.tsv:{number}: input case is absent"))
    })?;
    let same_identity = source.canonical_binary == identity.canonical_binary;
    if operation == "verify-claimed-content" && same_identity {
        return Err(GoldenError::violation(format!(
            "steps.tsv:{number}: mismatch uses matching content"
        )));
    }
    if operation != "verify-claimed-content" && !same_identity {
        return Err(GoldenError::violation(format!(
            "steps.tsv:{number}: exact operation substitutes identity"
        )));
    }
    if operation == "admit-exact" {
        admitted.insert(identity.canonical_binary.clone());
    }
    if operation == "read-exact" && !admitted.contains(&identity.canonical_binary) {
        return Err(GoldenError::violation(format!(
            "steps.tsv:{number}: exact read precedes admission"
        )));
    }
    Ok(())
}

fn step_outcome(operation: &str) -> Option<&'static str> {
    match operation {
        "identify" => Some("keep.identity.identified"),
        "admit-exact" => Some("keep.content.admitted"),
        "read-exact" => Some("keep.content.exact"),
        "verify-claimed-content" => Some("keep.content.mismatch"),
        "read-absent" => Some("keep.content.absent"),
        _ => None,
    }
}
